### PR TITLE
Network API

### PR DESCRIPTION
**Pull request by michaelfig**
Refs #748, #807 

This PR will be the SwingSet implementation of the new Network API.

### CODE REVIEW

**Comment by warner** [comment]
@tyg @michaelfig I have a preliminary writeup of the new "networking API" here:

https://github.com/Agoric/agoric-sdk/blob/mfig/channels/packages/SwingSet/docs/networking.md

Michael: you'll probably have to clean up some inconsistencies that I ran into (marked with TODO) before Tom can start the wordsmithing. I don't know that we have good answers for everything.. some of it is a matter of design decisions we still need to make. But some is deeper (the synchronous-ish ACK data in particular).

**Comment by michaelfig** [comment]
> Michael: you'll probably have to clean up some inconsistencies that I ran into (marked with TODO) before Tom can start the wordsmithing. I don't know that we have good answers for everything.. some of it is a matter of design decisions we still need to make. But some is deeper (the synchronous-ish ACK data in particular).

I've done my best.  The main change was renaming Channel to Connection in all the sources, then speaking of IBC `Relayer Connection`.  Could be improved, but I'm going to bed.

As far as ACK, I've explained how I want it to work, and then it depends on the particular Interface (renamed from Peer) to decide how to do it.  Basically, empty ACKs are the default, but if you're careful, you can use data-ful acks with certain interfaces.  This allows us to implement ICS-20-transfer as-is, with a hope to improve/clarify the IBC ACK definition in the future.

File: packages/SwingSet/src/vats/network/network.js [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+ * Convert some data to bytes.
+ *
+ * @param {Data} data
+ * @returns {Bytes}
+ */
+export function toBytes(data) {
+  // We really need marshallable TypedArrays.
+  if (typeof data === 'string') {
+    // eslint-disable-next-line no-bitwise
+    data = data.split('').map(c => c.charCodeAt(0));
```
Comment (warner) [inline packages/SwingSet/src/vats/network/network.js:93]: eww.. is that to avoid relying on Node's `Buffer` type to do the encoding? I suppose I'm ok with that as a goal, but I expect we'll fix `marshal` to handle bytes before we get the rest of this working in a non-Node environment.

(oh, but, I'm not even sure we provide `Buffer` inside SES compartments, so maybe this is the only way to go)

File: packages/cosmic-swingset/lib/ag-solo/vats/bootstrap.js [hunk 2]
```diff
@@ -69,16 +76,18 @@ export default function setup(syscall, state, helpers) {
...
         );
-        const zoe = await E(vats.zoe).getZoe();
-        const contractHost = await E(vats.host).makeHost();
 
         // Make the other demo mints
         const issuerNames = ['moola', 'simolean'];
-        const issuers = await Promise.all(
-          issuerNames.map(issuerName =>
-            E(vats.mints).makeMintAndIssuer(issuerName),
+        const [zoe, contractHost, issuers] = await Promise.all([
```
Comment (warner) [inline packages/cosmic-swingset/lib/ag-solo/vats/bootstrap.js]: why is this better? it's not like we have any meaningful parallelism within the kernel, and everything here is local, and it requires the reader to remember/compare the order of items in `[zoe, contractHost, issuers]` against the order of the inputs to `Promise.all`, whereas the original did exactly one thing per line. And this is unrelated to the networking changes.

File: packages/cosmic-swingset/lib/ag-solo/vats/bootstrap.js [hunk 3]
```diff
@@ -116,6 +129,47 @@ export default function setup(syscall, state, helpers) {
...
+        // Every vat has a loopback device.
+        ps.push(
+          E(networkVat).registerInterfaceHandler(
+            '/local',
+            makeLoopbackInterfaceHandler(),
+          ),
+        );
+        if (false && bridgeMgr) {
+          // TODO: Implement when off-chain, too!
+          // We have access to the bridge, and therefore IBC.
```
Comment (warner) [inline packages/cosmic-swingset/lib/ag-solo/vats/bootstrap.js:141]: hah, there's a lot of work hiding behind that claim :)

File: packages/produce-promise/src/producePromise.js [hunk 4]
```diff
@@ -1,12 +1,22 @@
+// @ts-check
```
Comment (warner) [inline packages/produce-promise/src/producePromise.js:1]: lovely but better in a separate PR

File: packages/store/src/store.js [hunk 5]
```diff
@@ -1,14 +1,33 @@
 // Copyright (C) 2019 Agoric, under Apache license 2.0
+// @ts-check
```
Comment (warner) [inline packages/store/src/store.js:2]: also great for a separate PR

**Review by warner — APPROVED** [review]
Looks good enough to land. For next time, I'd love it if the `SwingSet` changes were in one PR (adding a new exported feature), the `cosmic-swingset` changes which use that feature went into another, and all the unrelated cleanups went into a third, so I could review one thing at a time. But in the interests of expediency, I'm ok with landing all of this in a single PR.

File: packages/cosmic-swingset/lib/ag-solo/vats/bootstrap.js (hunk 2 above)
Comment (michaelfig) [reply packages/cosmic-swingset/lib/ag-solo/vats/bootstrap.js]: I'll revert.  I was doing something more in that change before.

File: packages/SwingSet/src/vats/network/network.js (hunk 1 above)
Comment (michaelfig) [reply packages/SwingSet/src/vats/network/network.js:93]: Yup.  And the hack is documented in `networking.md`.
